### ISSUE TITLE
SDRplay API 3.06 build errors

### ISSUE THREAD

**Issue by jhancuff** [fventuri/SoapySDRPlay3#1]
Greets,

Trying to build this driver against API 3.06.  Getting the following build errors.  Did SDR play change up some of the functions in their API since 3.01 to the point that your code is expecting different members?

`Scanning dependencies of target sdrPlay3Support
[ 20%] Building CXX object CMakeFiles/sdrPlay3Support.dir/Registration.cpp.o
[ 40%] Building CXX object CMakeFiles/sdrPlay3Support.dir/Settings.cpp.o
/home/pi/SoapySDRPlay3/Settings.cpp: In constructor ‘SoapySDRPlay3::SoapySDRPlay3(const Kwargs&)’:
/home/pi/SoapySDRPlay3/Settings.cpp:149:30: error: ‘struct sdrplay_api_AgcT’ has no member named ‘knee_dBfs’
     chParams->ctrlParams.agc.knee_dBfs = 0;
                              ^
/home/pi/SoapySDRPlay3/Settings.cpp:152:30: error: ‘struct sdrplay_api_AgcT’ has no member named ‘hang_ms’
     chParams->ctrlParams.agc.hang_ms = 0;
                              ^
/home/pi/SoapySDRPlay3/Settings.cpp:154:30: error: ‘struct sdrplay_api_AgcT’ has no member named ‘LNAstate’
     chParams->ctrlParams.agc.LNAstate = chParams->tunerParams.gain.LNAstate;
                              ^
/home/pi/SoapySDRPlay3/Settings.cpp: In member function ‘virtual void SoapySDRPlay3::setAntenna(int, size_t, const string&)’:
/home/pi/SoapySDRPlay3/Settings.cpp:294:98: error: too few arguments to function ‘sdrplay_api_ErrT sdrplay_api_Update(HANDLE, sdrplay_api_TunerSelectT, sdrplay_api_ReasonForUpdateT, sdrplay_api_ReasonForUpdateExtension1T)’
                 sdrplay_api_Update(device.dev, device.tuner, sdrplay_api_Update_Rsp2_AmPortSelect);
                                                                                                  ^
In file included from /home/pi/SoapySDRPlay3/SoapySDRPlay3.hpp:40:0,
                 from /home/pi/SoapySDRPlay3/Settings.cpp:27:
`

It then goes on for some time with errors such as 
`In file included from /home/pi/SoapySDRPlay3/SoapySDRPlay3.hpp:40:0,
                 from /home/pi/SoapySDRPlay3/Settings.cpp:27:
/usr/local/include/sdrplay_api.h:216:52: note: declared here
     _SDRPLAY_DLL_QUALIFIER sdrplay_api_ErrT        sdrplay_api_Update(HANDLE dev, sdrplay_api_TunerSelectT tuner, sdrplay_api_ReasonForUpdateT reasonForUpdate, sdrplay_api_ReasonForUpdateExtension1T re
                                                    ^
/home/pi/SoapySDRPlay3/Settings.cpp:359:96: error: too few arguments to function ‘sdrplay_api_ErrT sdrplay_api_Update(HANDLE, sdrplay_api_TunerSelectT, sdrplay_api_ReasonForUpdateT, sdrplay_api_ReasonForUpdateExtension1T)’
             sdrplay_api_Update(device.dev, device.tuner, sdrplay_api_Update_RspDuo_AmPortSelect);
                                                                                                ^
`
which I imagine is largely all getting set off from the first error I posted (which is the first viewable error.  Lemme know if you want the entire error output.

My ultimate goal is to get SoapyRemote working with the RSPdx as my main listening rig is an RPi 4 in my attic with all of the nice antennas connected to it.

**Comment by SDRplay** [comment]
Yes the API 3 has had some significant improvements since it's first release. If you are building against the 3.06 API then you must have the 3.06 include files? Your mention of 3.01 suggests that you have out of date files?

Can you confirm what API binary and include files you have installed and are building against?

**Comment by SDRplay** [comment]
I've just realised you are building from the wrong repo. Try this...

git clone -b API3+RSPduo https://github.com/fventuri/SoapySDRPlay

This issue can be closed now.

**Comment by jhancuff** [comment]
I'm building against ARM32 API 3.06.1, just tried building from the repo API3-RSPduo branch and received the following build error:
`Scanning dependencies of target sdrPlaySupport
[ 20%] Building CXX object CMakeFiles/sdrPlaySupport.dir/Registration.cpp.o
[ 40%] Building CXX object CMakeFiles/sdrPlaySupport.dir/Settings.cpp.o
[ 60%] Building CXX object CMakeFiles/sdrPlaySupport.dir/Streaming.cpp.o
/home/pi/SoapySDRPlay/Streaming.cpp: In member function ‘virtual int SoapySDRPlay::activateStream(SoapySDR::Stream*, int, long long int, size_t)’:
/home/pi/SoapySDRPlay/Streaming.cpp:302:49: error: ‘sdrplay_api_SetTransferMode’ was not declared in this scope
     sdrplay_api_SetTransferMode(sdrplay_api_BULK);
                                                 ^
make[2]: *** [CMakeFiles/sdrPlaySupport.dir/build.make:89: CMakeFiles/sdrPlaySupport.dir/Streaming.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/sdrPlaySupport.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
[ 20%] Building CXX object CMakeFiles/sdrPlaySupport.dir/Streaming.cpp.o
/home/pi/SoapySDRPlay/Streaming.cpp: In member function ‘virtual int SoapySDRPlay::activateStream(SoapySDR::Stream*, int, long long int, size_t)’:
/home/pi/SoapySDRPlay/Streaming.cpp:302:49: error: ‘sdrplay_api_SetTransferMode’ was not declared in this scope
     sdrplay_api_SetTransferMode(sdrplay_api_BULK);
                                                 ^
make[2]: *** [CMakeFiles/sdrPlaySupport.dir/build.make:89: CMakeFiles/sdrPlaySupport.dir/Streaming.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/sdrPlaySupport.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
`

**Comment by jhancuff** [comment]
I had mentioned the 3.01 version as a reference to the first version you original built for (at least that I could see that someone compiled for.  Wasn't sure if you'd be updating this repo as the API evolved.  I've only ever had the latest version (as of today) installed along with the latest include files.  I actually scoured my system to ensure I didn't have any latent sdrplay libs or includes on my RPi.

**Comment by SDRplay** [comment]
Delete lines 300-303 and try again - I'll need to discuss this with Franco - those lines are no longer needed.

Sorry for the confusion,

Andy

**Comment by jhancuff** [comment]
I was just about to suggest that actually.  Compiled fine after removing the undeclared function.  Thanks for the super quick feedback!

-Joe

**Comment by fventuri** [comment]
Thanks for the input, Andy and Joe.

I just pushed a commit to the API3+RSPduo branch in the SoapySDRPlay repo that removes those three lines (my desktop here is a x86_64 architecture, so my builds never saw that invalid API function call).
I also added a singleton object called 'sdrplay_api', that takes care of opening/closing the SDRplay API using RAII; this way the code is cleaner and sdrplay_api_Open()/sdrplay_api_Close() are invoked exactly once when the client ('CubicSDR') uses the SoapySDRPlay driver.

Let me know if you find any problem with these changes,
Franco
